src/sync/partial_sync.cpp

```diff
@@ -56,6 +151,123 @@ void update_schema(Group& group, Property matches_property)
         ObjectStore::apply_additive_changes(group, required_changes, true);
 }
 
+struct ResultSetsColumns {
+    ResultSetsColumns(Table& table, std::string const& matches_property_name)
+    {
+        name = table.get_column_index("name");
+        REALM_ASSERT(name != npos);
+
+        query = table.get_column_index("query");
+        REALM_ASSERT(query != npos);
+
+        this->matches_property_name = table.get_column_index("matches_property");
+        REALM_ASSERT(this->matches_property_name != npos);
+
+        // This may be `npos` if the column does not yet exist.
+        matches_property = table.get_column_index(matches_property_name);
+    }
+
+    size_t name;
+    size_t query;
+    size_t matches_property_name;
+    size_t matches_property;
+};
+
+bool validate_existing_subscription(Table& table, ResultSetsColumns const& columns, std::string const& name,
+                                    std::string const& query, std::string const& matches_property)
+{
+    auto existing_row_ndx = table.find_first_string(columns.name, name);
+    if (existing_row_ndx == npos)
+        return false;
+
+    StringData existing_query = table.get_string(columns.query, existing_row_ndx);
+    if (existing_query != query)
+        throw std::runtime_error("An existing subscription exists with the same name, but a different query.");
```

Comment:
Putting the non-matching queries and result types in these errors could aid debugging.